x-pack/plugins/triggers_actions_ui/public/application/sections/alerts_search_bar/url_synced_alerts_search_bar.tsx

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useCallback, useEffect } from 'react';
+import { BoolQuery } from '@kbn/es-query';
+import { AlertConsumers } from '@kbn/rule-data-utils';
+import { AlertsFeatureIdsFilter } from '../../lib/search_filters';
+import { useKibana } from '../../..';
+import { useAlertSearchBarStateContainer } from './use_alert_search_bar_state_container';
+import { ALERTS_URL_STORAGE_KEY } from './constants';
+import { AlertsSearchBarProps } from './types';
+import AlertsSearchBar from './alerts_search_bar';
+import { buildEsQuery } from '../global_alerts_page/global_alerts_page';
+import { nonNullable } from '../../../../common/utils';
+
+export type UrlSyncedAlertsSearchBarProps = Omit<
+  AlertsSearchBarProps,
+  'query' | 'rangeFrom' | 'rangeTo' | 'filters' | 'onQuerySubmit'
+> & {
+  onEsQueryChange: (esQuery: { bool: BoolQuery }) => void;
+  onActiveFeatureFiltersChange?: (value: AlertConsumers[]) => void;
+};
+
+/**
+ * An abstraction over AlertsSearchBar that syncs the query state with the url
+ */
+export const UrlSyncedAlertsSearchBar = ({
+  onEsQueryChange,
+  onActiveFeatureFiltersChange,
+  ...rest
+}: UrlSyncedAlertsSearchBarProps) => {
+  const {
+    data: { query: queryService },
+  } = useKibana().services;
+  const {
+    timefilter: { timefilter: timeFilterService },
+  } = queryService;
+  const {
+    kuery,
+    rangeFrom,
+    rangeTo,
+    filters,
+    onKueryChange,
+    onRangeFromChange,
+    onRangeToChange,
+    onFiltersChange,
+    savedQuery,
+    setSavedQuery,
+    clearSavedQuery,
+  } = useAlertSearchBarStateContainer(ALERTS_URL_STORAGE_KEY);
+
+  const syncEsQuery = useCallback(() => {
+    try {
+      onActiveFeatureFiltersChange?.([
+        ...new Set(
+          filters
+            .flatMap((f) => (f as AlertsFeatureIdsFilter).meta.alertsFeatureIds)
+            .filter(nonNullable)
+        ),
+      ]);
+      const newQuery = buildEsQuery({
+        timeRange: {
+          to: rangeTo,
+          from: rangeFrom,
+        },
+        kuery,
+        filters,
+      });
+      onEsQueryChange(newQuery);
+    } catch (e) {
+      // TODO show error message?
+    }
+  }, [filters, kuery, onActiveFeatureFiltersChange, onEsQueryChange, rangeFrom, rangeTo]);
+
+  useEffect(() => {
```

Comment:
Interesting and clever to trigger an effect when a `useCallback`'s dependencies change, but I'm wary of it. React has been threatening to remove manual dependency definitions from `useEffect` almost since they introduced it, and yeah, it's probably vaporware, but it's still a pattern we should be careful with.

What's the expected behavior? When are we expecting `syncEsQuery` to be called again as a side-effect of a re-render? I can see it's being called in `onQueryChange`, which makes sense, but reading the code I can't tell when I should expect this effect to trip.

If there isn't a way of achieving this without `useEffect` can you add a comment to explain it?

Reply:
You made me realize that there was a much less intricate and more readable way to trigger these updates: I removed the callback and switched a normal `useEffect`, thanks! 🙂